### PR TITLE
refactor: Change replay back mechanism

### LINKED ISSUE #97: Consider to change replay back mechanism
### File Location(s)

_No response_

### Proposal

The current implementation of the replay back is straightforward: when running the relayer, it first checks whether the algorithm needs to be run. If the test result is positive, then the replay back starts and then the program exits. Thus requiring a repeated restart.

Modification involves continuing the relayer after the missing blocks has been replayed back. A gear program change will be required.

If the replay back is required in the main loop, the program stops.

### PR DESCRIPTION
**Pull request by gshep**
Resolves #97 

- continue execution after blocks have been replayed back
- create gear events listener on demand to avoid errors spam on pruning nodes
- 

@gear-tech/dev

### CODE REVIEW

File: Cargo.toml
Comment (mertwole) [inline Cargo.toml:1]: Please use versions and toolchain from https://github.com/gear-tech/gear-bridges/pull/93/commits/038cc3bef427d0a399887b66bfa21da81c5db2ee as we need to stick to the version that will be deployed on testnet

File: relayer/src/ethereum_checkpoints/mod.rs [hunk 1]
```diff
@@ -186,10 +168,15 @@ pub async fn relay(args: RelayCheckpointsArgs) {
                 }
             }
             Ok(Err(sync_update::Error::ReplayBackRequired { .. })) => {
-                log::info!("Replay back within the main loop. Exiting");
+                log::error!("Replay back within the main loop. Exiting");
                 return;
             }
-            Ok(Err(e)) => log::info!("The program failed with: {e:?}. Skipping"),
+            Ok(Err(e)) => {
+                log::info!("The program failed with: {e:?}. Skipping");
```
Comment (mertwole) [inline relayer/src/ethereum_checkpoints/mod.rs:175]: Maybe `log::error!`?

File: relayer/src/ethereum_checkpoints/mod.rs [hunk 2]
```diff
@@ -186,10 +168,15 @@ pub async fn relay(args: RelayCheckpointsArgs) {
...
+                if let sync_update::Error::NotActual = e {
+                    slot_last = slot;
+                }
```
Comment (mertwole) [inline relayer/src/ethereum_checkpoints/mod.rs:178]: Why not a separate case in `match`?

**Review by mertwole — APPROVED** [review]

File: Cargo.toml
Comment (mertwole) [reply Cargo.toml:1]: now merged to `main`

File: relayer/src/ethereum_checkpoints/mod.rs (hunk 2 above)
Comment (gshep) [reply relayer/src/ethereum_checkpoints/mod.rs:178]: not to duplicate log record

Comment (mertwole) [reply relayer/src/ethereum_checkpoints/mod.rs:178]: But the case when this error appears is not a error, as I see, more like expected behaviour

Comment (gshep) [reply relayer/src/ethereum_checkpoints/mod.rs:178]: yep, just like others from this arm

Comment (mertwole) [reply relayer/src/ethereum_checkpoints/mod.rs:178]: Why? When the error don't match specific variant it seems like unexpected behaviour

**Pull request merged**